xctool should consider build commands that emit "error:" messages to have failed.

In Xcode, if you have a Run Script step that simply runs ...

```
echo "error: nothing really failed"
```

... _xcodebuild_ will consider that Run Script to have failed, even though it didn't emit a non-zero exit status.  It will ultimately print "BUILD FAILED" at the end, and exit with a non-zero exit status.

When this happens, xctool should consider this to be a failed build command.

e.g., right now, we'd get the following output from xctool ...

```
=== BUILD ===

  xcodebuild build SomeProject
    SomeProject / SomeProject (Debug)
      ✓ Check dependencies (32 ms)
      ✓ Run custom shell script 'Run Script' (12 ms)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
error: nothing really failed.
Command /bin/sh emitted errors but did not return a nonzero exit code to indicate failure
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

** BUILD FAILED ** (462 ms)
```

**but**, we should get ...

```
=== BUILD ===

  xcodebuild build SomeProject
    SomeProject / SomeProject (Debug)
      ✓ Check dependencies (31 ms)
      ✗ Run custom shell script 'Run Script' (13 ms)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
PhaseScriptExecution Run\ Script /Users/fpotter/Library/Developer/Xcode/DerivedData/RegexTest-bpuwquewovhzwofkadpnppmxgfxn/Build/Intermediates/RegexTest.build/Debug/RegexTest.build/Script-285E35D218453321009E9BFD.sh
    cd /Users/fpotter/Desktop/RegexTest
    /bin/sh -c /Users/fpotter/Library/Developer/Xcode/DerivedData/RegexTest-bpuwquewovhzwofkadpnppmxgfxn/Build/Intermediates/RegexTest.build/Debug/RegexTest.build/Script-285E35D218453321009E9BFD.sh

error: nothing really failed.
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━


Failures:

  0) Run custom shell script 'Run Script'
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
PhaseScriptExecution Run\ Script /Users/fpotter/Library/Developer/Xcode/DerivedData/RegexTest-bpuwquewovhzwofkadpnppmxgfxn/Build/Intermediates/RegexTest.build/Debug/RegexTest.build/Script-285E35D218453321009E9BFD.sh
    cd /Users/fpotter/Desktop/RegexTest
    /bin/sh -c /Users/fpotter/Library/Developer/Xcode/DerivedData/RegexTest-bpuwquewovhzwofkadpnppmxgfxn/Build/Intermediates/RegexTest.build/Debug/RegexTest.build/Script-285E35D218453321009E9BFD.sh

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

** BUILD FAILED ** (477 ms)
```

--- a/Common/ReporterEvents.h
+++ b/Common/ReporterEvents.h
@@ -95,6 +95,9 @@
 #define kReporter_EndBuildCommand_SucceededKey @"succeeded"
 #define kReporter_EndBuildCommand_EmittedOutputTextKey @"emittedOutputText"
 #define kReporter_EndBuildCommand_DurationKey @"duration"
+#define kReporter_EndBuildCommand_ResultCode @"resultCode"
+#define kReporter_EndBuildCommand_TotalNumberOfWarnings @"totalNumberOfWarnings"
+#define kReporter_EndBuildCommand_TotalNumberOfErrors @"totalNumberOfErrors"
 
 #define kReporter_BeginBuildTarget_ProjectKey @"project"
 #define kReporter_BeginBuildTarget_TargetKey @"target"

--- a/reporters/text/TextReporter.m
+++ b/reporters/text/TextReporter.m
@@ -530,7 +530,7 @@ static NSString *abbreviatePath(NSString *string) {
 
   NSString *indicator = nil;
   if (succeeded) {
-    if ([outputText rangeOfString:@"warning:"].location != NSNotFound) {
+    if ([event[kReporter_EndBuildCommand_TotalNumberOfWarnings] unsignedIntegerValue] > 0) {
       indicator = [self warningIndicatorString];
     } else {
       indicator = [self passIndicatorString];

--- a/xcodebuild-shim/xcodebuild-shim/xcodebuild_shim.m
+++ b/xcodebuild-shim/xcodebuild-shim/xcodebuild_shim.m
@@ -43,6 +43,7 @@ static void XTSwizzleSelectorForFunction(Class cls, SEL sel, IMP newImp)
   method_exchangeImplementations(originalMethod, newMethod);
 }
 
+// from IDEFoundation.framework
 @interface IDEActivityLogSection : NSObject
 
 // Will be 0 for success, 1 for failure.
@@ -69,6 +70,12 @@ static void XTSwizzleSelectorForFunction(Class cls, SEL sel, IMP newImp)
 
 @property(readonly) double timeStoppedRecording;
 @property(readonly) double timeStartedRecording;
+
+// The number of times "warning:" appears in the output.
+@property(readonly) NSUInteger totalNumberOfWarnings;
+
+// The number of times "error:" appears in the output.
+@property(readonly) NSUInteger totalNumberOfErrors;
 
 @end
 
@@ -170,12 +177,18 @@ static void AnnounceEndSection(IDEActivityLogSection *section)
     PrintJSON(EventDictionaryWithNameAndContent(
       kReporter_Events_EndBuildCommand, @{
         kReporter_EndBuildCommand_TitleKey : section.title,
-        kReporter_EndBuildCommand_SucceededKey : (section.resultCode == 0) ? @YES : @NO,
+        // Xcode will only consider something a success if resultCode == 0 AND
+        // there's no output that contains the string "error: ".
+        kReporter_EndBuildCommand_SucceededKey : ((section.resultCode == 0) &&
+                                                  (section.totalNumberOfErrors == 0)) ? @YES : @NO,
         // Sometimes things will fail and 'emittedOutputText' will be nil.  We've seen this
         // happen when Xcode's Copy command fails.  In this case, just send an empty string
         // so Reporters don't have to worry about this sometimes being [NSNull null].
         kReporter_EndBuildCommand_EmittedOutputTextKey : section.emittedOutputText ?: @"",
         kReporter_EndBuildCommand_DurationKey : @(section.timeStoppedRecording - section.timeStartedRecording),
+        kReporter_EndBuildCommand_ResultCode : @(section.resultCode),
+        kReporter_EndBuildCommand_TotalNumberOfWarnings : @(section.totalNumberOfWarnings),
+        kReporter_EndBuildCommand_TotalNumberOfErrors : @(section.totalNumberOfErrors),
       }));
   } else if ([sectionTypeString hasPrefix:kDomainTypeProductItemPrefix]) {
     NSString *project = nil;

